### PR TITLE
feat: support externally hosted plugins

### PR DESCRIPTION
**Pull request by PhilippLgh**
#### What does it do?

It extends the PluginHost with functionality to process a list `plugins.json` of externally hosted plugins and allows to download & cache them, verify their signatures and eventually load them in the main process during runtime. These plugins are stored in a path outside of the main app so that changes to them don't require elevated user permissions or re-installation.

#### Any helpful background information?

The purpose of this PR is to give external projects more control and autonomy over their integrations and the release lifecycle. Moreover, changes to external plugins won't affect Grid's release cycle since such plugins can be updated, removed and reloaded during runtime.

*Problem*: the renderer start is deferred until *all* plugins (local & remote) are processed. This is mostly done because otherwise it would require changes in the state handling in grid-ui. An event hook `plugins-loaded` was added to allow the UI to listen for state changes.

#### New dependencies? What are they used for?

It will require a newer version of electron-app-manager that includes the verification info in the plugin's release metadata.

### CODE REVIEW

**Comment by marcgarreau** [comment]
Notes make mention of a new version of `electron-app-manager`, but none included in the PR. Is that required to run this changeset or a future iteration?

File: ethereum_clients/PluginHost.js [hunk 1]
```diff
@@ -1,34 +1,158 @@
...
+    }
+    let releases = remotePluginList.map(async pluginShortInfo => {
+      try {
+        const { name: pluginName, location } = pluginShortInfo
+        if (!location) {
+          throw new Error(
+            `error: external plugin ${pluginName} does not specify a valid location`
+          )
+        }
+        if (fs.existsSync(location)) {
```
Comment (evertonfraga) [inline ethereum_clients/PluginHost.js]: while testing, i couldn't have Raiden plugin to appear. On a debug session, turned out `location` was returning the URL. Is that expected?


image

**Comment by evertonfraga** [comment]
@PhilippLgh alright, I resumed the debugging session. Some comments:

With a cleared plugins directory, I get [this thrown error](https://github.com/ethereum/grid/pull/202/files#diff-6aa09c4df2fdba3a219c60d888d6e6d5R109) right after `raiden` plugin is downloaded. It seems `pluginManager.download(latest)` does not perform the checks that `pluginManager.getLatest()` does. The former does not have `verificationResults` property.

On a second run it worked, because the plugin was already downloaded and `getLatest()` seems to perform the checks.

File: ethereum_clients/PluginHost.js [hunk 2]
```diff
@@ -1,34 +1,158 @@
...
+      )
+    } catch (error) {
+      console.log('error: could not parse plugin list', error)
+    }
+    let releases = remotePluginList.map(async pluginShortInfo => {
+      try {
+        const { name: pluginName, location } = pluginShortInfo
+        if (!location) {
+          throw new Error(
+            `error: external plugin ${pluginName} does not specify a valid location`
```
Comment (evertonfraga) [inline ethereum_clients/PluginHost.js]: I think we should perform a validation over whole plugin schema here, so we don't get unexpected results down the road.

File: ethereum_clients/PluginHost.js (hunk 1 above)
Comment (PhilippLgh) [reply ethereum_clients/PluginHost.js]: location can be local path or url

File: ethereum_clients/PluginHost.js (hunk 2 above)
Comment (PhilippLgh) [reply ethereum_clients/PluginHost.js]: good catch. fix is provided through new app-manager version

**Pull request merged**